python/graphframes/examples.py

```diff
@@ -57,3 +60,216 @@ def friends(self):
+    Thin Junction Trees." AISTATS, 2011.
+
+    The BP algorithm works by:
+
+    * Coloring the graph by assigning a color to each vertex such that no neighboring vertices
+      share the same color.
+    * In each step of BP, update all vertices of a single color.  Alternate colors.
+     """
+    @classmethod
+    def main(cls):
+        """Run the algorithm for an example problem."""
+        conf = SparkConf().setAppName("BeliefPropagation example")
+        sc = SparkContext.getOrCreate(conf)
+        sql = SQLContext.getOrCreate(sc)
+
+        # create graphical model g of size 3 x 3
+        g = Graphs(sql).gridIsingModel(3)
+        print("Original Ising model:")
+        g.vertices.show()
+        g.edges.show()
+
+        # run BP for 5 iterations
+        numIter = 5
+        results = cls.runBPwithGraphFrames(g, numIter)
+
+        # display beliefs
+        beliefs = results.vertices.select('id', 'belief')
+        print("Done with BP. Final beliefs after {} iterations:".format(numIter))
+        beliefs.show()
+
+        sc.stop()
+
+    @staticmethod
+    def colorGraph(g):
+        """Given a GraphFrame, choose colors for each vertex.
+
+        No neighboring vertices will share the same color. The number of colors is minimized.
+
+        This is written specifically for grid graphs. For non-grid graphs, it should be generalized,
+        such as by using a greedy coloring scheme.
+
+        :param g: Grid graph generated by :meth:`Graphs.gridIsingModel()`
+        :return: Same graph, but with a new vertex column "color" of type Int (0 or 1)
+
+        """
+
+        colorUDF = sqlFuncs.udf(
+            lambda i, j: 0 if (i + j) % 2 == 0 else 1,
+            returnType=types.IntegerType())
+        v = g.vertices.withColumn('color', colorUDF(sqlFuncs.col('i'), sqlFuncs.col('j')))
+        return GraphFrame(v, g.edges)
+
+    @classmethod
+    def runBPwithGraphFrames(cls, g, numIter):
+        """Run Belief Propagation using GraphFrame.
+
+        This implementation of BP shows how to use GraphFrame's aggregateMessages method.
+        """
+        # choose colors for vertices for BP scheduling
+        colorG = cls.colorGraph(g)
+        numColors = colorG.vertices.select('color').distinct().count()
+
+        # TODO: handle vertices without any edges
+
+        # initialize vertex beliefs at 0.0
+        gx = GraphFrame(colorG.vertices.withColumn('belief', sqlFuncs.lit(0.0)), colorG.edges)
+
+        # run BP for numIter iterations
+        for iter_ in range(numIter):
+            # for each color, have that color receive messages from neighbors
+            for color in range(numColors):
+                # Send messages to vertices of the current color.
+                # We may send to source or destination since edges are treated as undirected.
+                msgForSrc = sqlFuncs.when(
+                    AM.src()['color'] == color,
+                    AM.edge()['b'] * AM.dst()['belief'])
```

Comment:
I thought about this also.

In Scala we can call function `c = AM.src` without parenthesis and get a `Column` instance. Furthermore, the returned `Column` instance is indexed with `c2 = c('color')`.

In Python functions are not (typically) called without parenthesis so we need to do `c = AM.src()`. Also the indexing is done with `c2 = c['color']`.

The empty parenthesis in Python can be dropped out with a hack but this is not conventional. Class `AM` is just for grouping purposes to make the structure consistent with Scala and it is not initialised. If `AM` would be initialised, `@property` decorator could be used to allow calling `AM.src` without parenthesis. I can make some custom modifications to allow applying similar decorator to create a "class property". However, I believe the concept of properties in python is meant for different purpose and it would become confusing. I would not recommend it.

What do you think? Did I get something wrong?